channeldb/invoices.go

```diff
@@ -1879,11 +1887,195 @@ func settleHtlcsAmp(invoice *invpkg.Invoice,
 	updateMap[setID][circuitKey] = htlc
 }
 
+// InvoiceUpdaterContext is the context that's passed to the methods of the
+// InvoiceUpdater interface. It contains all the information needed to update
+// the invoice in the database.
+type InvoiceUpdaterContext struct {
+	// invoice holds the updated invoice. It is always set regardless
+	// of the type of update.
+	invoice *invpkg.Invoice
+
+	// updateTime is the timestamp used for the update.
+	updateTime time.Time
+
+	// settledSetIDs holds the set of set IDs that were settled as part of
+	// this update. This is only set for StoreAddHtlcsUpdate.
+	settledSetIDs map[invpkg.SetID]struct{}
+
+	// ampHtlcsUpdate holds the set of AMP HTLCs that were added or
+	// cancelled as part of this update. This is only set for
+	// StoreAddHtlcsUpdate and StoreCancelHtlcsUpdate.
+	ampHtlcsUpdate map[invpkg.SetID]map[models.CircuitKey]*invpkg.InvoiceHTLC
+}
+
+// InvoiceUpdater is an interface to abstract away the details of updating an
+// invoice in the database. The methods of this interface are called when the
+// in-memory update of an invoice is complete, and the database needs to be
+// updated.
+type InvoiceUpdater interface {
+	// StoreCancelHtlcsUpdate updates the invoice in the database after
+	// cancelling a set of HTLCs.
+	StoreCancelHtlcsUpdate(InvoiceUpdaterContext) error
```

Comment:
What about a slightly diff model where: 
  1. We have two main methods `ApplyUpdate`, then `Commit`. 
  2. `ApplyUpdate` takes something similar to `InvoiceUpdaterContext`, but specific for each type of update. This is a new interface, `InvoiceUpdate` which is then destructed within the `ApplyUpdate` method to figure. out how to apply it. 
  3. `Commit` makes the db transaction, and tries to commit it all. If this fails, everything else above fails as well. 

Eg we can take `StoreAddHtlcsUpdate` and make a new struct of `NewHtlcUpdate` that just has a field of the HTLCs to update, and w/e other information. This gets passed into `ApplyUpdate` that figures out how to stage it, if the state is invalid, etc. Ideally we can have the concrete implementation of the `*Update` structs handle the validation. Eg: you can't add more HTLCs if the state of the invoice is already cancelled, etc. Then we have all these smaller update modules, that can be tested in isolation. 

The main `InvoiceUpdate` interface can be pretty slim. Mainly just to allow polymorphism to simplify the call site at the invoice registry: figure out all updates, make a new updater, apply then all, then commit. So maybe: 
```go
type InvoiceUpdate interface {
      InvoiceRef() invoice.InvoiceRef
      Type() invoice.UpdateType
      ValidateUpdate(oldInvoice *Invoice) error 
}
```

Reply:
IUUC what you describe is essentially the same model that's present in the PR + the existing in-memory update methods, with the added benefit of more streamlines approach through types and interfaces. So in this proposed approach the "ApplyUpdate" part is where we apply the update to the copied invoice in memory (with the existing legacy methods) and "Commit" is the `InvoiceUpdater` interface without the nice abstraction since we pass around the mutated+new state in the context.

Reply:
I see a deeper refactor where the `invoices.Invoice` type could only be mutated through specific methods and the update itself would just call into those methods. That way the mutated state could just be queue'd up and committed to the database later. I'm open to such big change too, but its also not without risk as it implies changes to existing tests etc.

Reply:
I think your suggestion is very similar to what @ziggie1984 expressed below, but I also think that doing a full-fledged "deep" refactor would be a lot more work that I'd like to first justify vs a simpler approach presented here. I think a more gradual multi PR refactor could be built as follow up once the e2e functionality is rock solid. The approach presented in this PR has a working SQL implementation too which we could use (once it's solidified) as a v1. wdyt?